## Laboratorio 6: Análisis de tráfico de red

#### Brandon Ronaldo Sicay Cumes - 21757

### **Parte 1 – Filtrado y preprocesamiento**


1. Cargue la información del archivo large_even.json en una lista, muestre la cantidad de
registros total (deben ser 746, 909). Este es nuestro tráfico inicial!

In [40]:
import json
import pandas as pd
from pandas import json_normalize
import pickle

In [41]:

data = list()
with open("large_eve.json", "r") as file:
    for line in file:
        json_file = json.loads(line)
        data.append(json_file)

print("Cantidad total de registros:", len(data))

Cantidad total de registros: 746909


2. Debido a que estamos buscando dominios web, del total de registros, solamente estamos
interesados en los registros DNS. Cargue únicamente aquellos registros que sean DNS.

In [42]:
dns_data = [item for item in data if item["event_type"] == "dns"]


3. Muestre la nueva cantidad de registros filtrados. 

In [43]:
print("Cantidad de registros DNS filtrados:", len(dns_data))
dataframe = json_normalize(dns_data)



Cantidad de registros DNS filtrados: 15749


4. Muestre la información de 2 registros cualesquiera.

In [44]:
dataframe.head(2)


timestamp           flow_id  pcap_cnt event_type  \
0  2017-07-22T17:33:16.661646-0500  1327836194150542     22269        dns   
1  2017-07-22T17:33:24.990320-0500  2022925111925872     54352        dns   

   vlan                                   src_ip  src_port  \
0   110  2001:0dbb:0c18:0011:0260:6eff:fe30:0863     59680   
1   110  2001:0dbb:0c18:0011:0260:6eff:fe30:0863     38051   

                                   dest_ip  dest_port proto dns.type  dns.id  \
0  2001:0500:0001:0000:0000:0000:803f:0235         53   UDP    query   15529   
1  2001:0500:0003:0000:0000:0000:0000:0042         53   UDP    query   58278   

             dns.rrname dns.rrtype  dns.tx_id dns.rcode  dns.ttl dns.rdata  
0  api.wunderground.com          A        0.0       NaN      NaN       NaN  
1   stork79.dropbox.com          A        0.0       NaN      NaN       NaN

5. Debido a que la data consiste en estructuras JSON anidadas, utilice la característica
json_normalize para normalizar la información y asignarla en un dataframe. Muestre el shape
del dataframe,

In [45]:
dataframe.shape


(15749, 18)

6. Como estamos buscando dominios DGA, debemos filtrar los registros DNS para aquellos
registros tipo A (son aquellos que mantienen una dirección IP asociada a un dominio). Después
de filtrar debería obtener 2849 registros.


In [46]:
dga_data = dataframe.loc[dataframe["dns.rrtype"] == "A", :]

print("Cantidad de registros DGA filtrados:", len(dga_data))

Cantidad de registros DGA filtrados: 2849


7. Filtre los dominios únicos. Debe obtener 177 registros únicos

In [47]:
len(dga_data["dns.rrname"].unique())


177

8. Escriba una función que obtenga el TLD para un dominio. Por ejemplo, para
api.wunderground.com el TLD es wunderground.com, para
safebrowsing.clients.google.com.home, el TLD es home. Utilice un LLM para ayudarle a
escribir esta función, verifique que obtiene correctamente el TLD, incluya el prompt utilizado
en su notebook.

In [48]:
def get_tld(domain):
    if domain.endswith('.home'):
        return 'home'
    else:
        parts = domain.split('.')
        tld = '.'.join(parts[-2:])
        return tld

domain1 = "api.wunderground.com"
domain2 = "safebrowsing.clients.google.com.home"

print("TLD para", domain1, "es:", get_tld(domain1))
print("TLD para", domain2, "es:", get_tld(domain2))

TLD para api.wunderground.com es: wunderground.com
TLD para safebrowsing.clients.google.com.home es: home


Promt:

    Escribe una función en python que obtenga el TLD para un dominio. De esta forma: Para api.wunderground.com el TLD es wunderground.com,  y para safebrowsing.clients.google.com.home, el TLD es home 


9. Del dataframe de dominios únicos de tipo A, obtenga el TLD (top level domain) utilizando la
función anterior para crear una columna nueva llamada domain_tld, y elimine todas las demás
columnas.

In [49]:
dga_data["domain_tld"] = dga_data["dns.rrname"].apply(get_tld)

domain_tld_dataframe = dga_data[["domain_tld"]]

domain_tld_dataframe

/var/folders/cy/x57dh4ds2g123w7nj93wcf_c0000gn/T/ipykernel_71277/2524569969.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dga_data["domain_tld"] = dga_data["dns.rrname"].apply(get_tld)


domain_tld
0      wunderground.com
1           dropbox.com
2             aoltw.net
5      wunderground.com
6                  home
...                 ...
15713       comcast.net
15716    stayonline.net
15725         aoltw.net
15737          real.com
15743       mozilla.com

[2849 rows x 1 columns]

In [67]:
print(domain_tld_dataframe.head())


         domain_tld
0  wunderground.com
1       dropbox.com
2         aoltw.net
5  wunderground.com
6              home


### **Parte 2 – Data Science**


10. Utilice Gemini para clasificar los dominios como DGA (1) o legítimos (0)


In [50]:
import google.generativeai as genai


In [55]:
import pandas as pd
import ast
from tqdm import tqdm
import re

genai.configure(api_key='AIzaSyCjx1oWiu7AovjKm9UE_Nr7BF-m52mrjJg')




prompt = f"""
Clasifica el siguiente dominio como DGA (1) o legítimo (0). Un dominio generado por algoritmo (DGA) suele tener una estructura aleatoria o poco comprensible, a menudo con letras y números sin sentido. Un dominio legítimo suele ser reconocible, como "google.com" o "microsoft.net".

Dominio: {domain}
Respuesta (solo 0 o 1). Devuelve tu respuesta SOLO como una lista de tuplas como esta:
[("dominio1.com", 0), ("dominio2.com", 1)]
No expliques nada. Solo devuelve la lista.
"""


saf = [
    {"category": "HARM_CATEGORY_DANGEROUS_CONTENT", "threshold": "BLOCK_NONE"},
    {"category": "HARM_CATEGORY_HARASSMENT", "threshold": "BLOCK_NONE"},
    {"category": "HARM_CATEGORY_SEXUALLY_EXPLICIT", "threshold": "BLOCK_NONE"},
    {"category": "HARM_CATEGORY_HATE_SPEECH", "threshold": "BLOCK_NONE"},
]

        

def clasificar_dominio_dga(domains, batch_size=20):
    results = []

    for i in tqdm(range(0, len(domains), batch_size), desc="Clasificando dominios con Gemini"):
        batch = domains[i:i+batch_size]
        input_text = prompt + "\n" + "\n".join(batch)

        try:
            response = genai.GenerativeModel('gemini-1.5-flash-001').generate_content(input_text, safety_settings=saf)
            raw_text = response.text.strip()
            fixed_text = raw_text if raw_text.startswith("[") else "[" + raw_text + "]"
            parsed = ast.literal_eval(fixed_text)
            # Validar que sea lista de tuplas
            for item in parsed:
                if isinstance(item, tuple) and len(item) == 2:
                    results.append(item)
        except Exception as e:
            print(f"Error en el batch {i}-{i+batch_size}: {e}")
            print("Respuesta cruda:", response.text)

            # Fallback por regex si falla el parseo
            fallback_matches = re.findall(r'\(?["\']?([\w\.-]+)["\']?,\s*([01])\)?', response.text)
            for match in fallback_matches:
                results.append((match[0], int(match[1])))

    return results

In [68]:
# ==========================================
# Aplicar al dataframe domain_tld_dataframe
# ==========================================

# Extraer dominios únicos
unique_domains = domain_tld_dataframe['domain_tld'].drop_duplicates().tolist()

# Clasificar con Gemini en bloques
classified_domains = clasificar_dominio_dga(unique_domains)

# Crear DataFrame
dga_classified_df = pd.DataFrame(classified_domains, columns=['domain_tld', 'is_dga'])


dga_final_df = domain_tld_dataframe.merge(dga_classified_df, on='domain_tld', how='left')



Clasificando dominios con Gemini: 100%|██████████| 6/6 [00:05<00:00,  1.07it/s]


11. Filtre aquellos considerados como DGA (valor 1 ) y muéstrelos. Recuerde que los resultados
de los modelos pueden incluir falsos positivos y falsos negativos, por lo que no podemos
fiarnos por completo de esta clasificación y debemos seguir indagando.

In [70]:
# Filtrar dominios clasificados como DGA
dga_only = dga_classified_df[dga_classified_df['is_dga'] == 1]

# Mostrar resultados
print(f"\nCantidad de dominios clasificados como DGA: {len(dga_only)}")
display(dga_only)


Cantidad de dominios clasificados como DGA: 22


domain_tld  is_dga
2                    aoltw.net       1
3                         home       1
8               stayonline.net       1
9             22.110phpmyadmin       1
11             com.localdomain       1
13   110phpmyadmin.localdomain       1
15               ntkrnlpa.info       1
23                         cn.       1
35                        wpad       1
36                     com.lan       1
37                     206.56"       1
45                     26-27.0       1
55                facebook.net       1
80                          FL       1
82                     saruman       1
83                     21.1201       1
91       whitecell.localdomain       1
97        vtlfccmfxlkgifuf.com       1
99                      22.254       1
100                    21-28.0       1
101       ejfodfmfxlkgifuf.xyz       1
102                    22.201:       1

### **Parte 3 – Dominio experto**


12. Ahora ya tenemos un listado de dominios reducido y considerado como sospechoso, por lo
que debemos aplicar dominio experto para encontrar los verdaderos registros maliciosos.
Escriba una función que utilice la lista de un millón de TLD proporcionada en Canvas, y
devuelva 0 si el TLD se encuentra en la lista y 1 si no está. Utilice un LLM para crear dicha
función, verifique que no se carga la lista cada vez que se busca un TLD. Incluya el prompt en
su notebook

In [71]:
top_1m = pd.read_csv("top-1m.csv", header=None, names=["rank", "domain"])
top_1m.head()

rank                     domain
0     1                 google.com
1     2             www.google.com
2     3              microsoft.com
3     4         data.microsoft.com
4     5  events.data.microsoft.com

Promt:

    Puedes ayudarme a escribir una función en python  que reciba dominios TLD de un csv proporcionado con columnas de número de fila y el dominio.  Y que  devuelva 0 si el TLD se encuentra en la lista, y 1 si no está. 


In [73]:
tld_set = top_1m["domain"].tolist()

def expert_domain_check(tld, tld_list=tld_set):
    if tld in tld_list:
        return 0
    else:
        return 1


13. Utilice la función para determinar si los TLD se encuentran en dicha lista. Filtre aquellos que si
se encuentran. Después de eliminar duplicados. 

In [77]:

dga_only['not_top1m'] = dga_only['domain_tld'].apply(expert_domain_check)


dga_only

/var/folders/cy/x57dh4ds2g123w7nj93wcf_c0000gn/T/ipykernel_71277/3480734844.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  dga_only['not_top1m'] = dga_only['domain_tld'].apply(expert_domain_check)


domain_tld  is_dga  not_top1m
2                    aoltw.net       1          1
3                         home       1          1
8               stayonline.net       1          1
9             22.110phpmyadmin       1          1
11             com.localdomain       1          1
13   110phpmyadmin.localdomain       1          1
15               ntkrnlpa.info       1          1
23                         cn.       1          1
35                        wpad       1          1
36                     com.lan       1          1
37                     206.56"       1          1
45                     26-27.0       1          1
55                facebook.net       1          0
80                          FL       1          1
82                     saruman       1          1
83                     21.1201       1          1
91       whitecell.localdomain       1          1
97        vtlfccmfxlkgifuf.com       1          1
99                      22.254       1          1
100                    21-28.0       1          1
101       ejfodfmfxlkgifuf.xyz       1          1
102                    22.201:       1          1

In [78]:
dga_expert_filtered = dga_only[dga_only['not_top1m'] == 1]

dga_expert_unique = dga_expert_filtered.drop_duplicates(subset=['domain_tld'])

print("Cantidad de dominios únicos en la lista de un millón de TLDs:", len(dga_expert_unique))
dga_expert_unique


Cantidad de dominios únicos en la lista de un millón de TLDs: 21


domain_tld  is_dga  not_top1m
2                    aoltw.net       1          1
3                         home       1          1
8               stayonline.net       1          1
9             22.110phpmyadmin       1          1
11             com.localdomain       1          1
13   110phpmyadmin.localdomain       1          1
15               ntkrnlpa.info       1          1
23                         cn.       1          1
35                        wpad       1          1
36                     com.lan       1          1
37                     206.56"       1          1
45                     26-27.0       1          1
80                          FL       1          1
82                     saruman       1          1
83                     21.1201       1          1
91       whitecell.localdomain       1          1
97        vtlfccmfxlkgifuf.com       1          1
99                      22.254       1          1
100                    21-28.0       1          1
101       ejfodfmfxlkgifuf.xyz       1          1
102                    22.201:       1          1

14. Finalmente, para confirmar los dominios maliciosos podemos buscar la fecha de creación del
TLD. Cree una función qué en base al TLD, devuelva la fecha de creación de este. Utilice un
LLM para escribir dicha función, incluya el prompt utilizado en su notebook.

Promt:

    Puedes ayudarme a escribir una función en python  que reciba dominios y en base a este, retorne la fecha de creación.  


In [97]:

def get_creation_date(tld):
    try:
        # Obtener información de whois para el TLD
        domain_info = whois.whois(tld)
        # Extraer la fecha de creación del TLD
        creation_date = domain_info.creation_date
        # Si creation_date es una lista, tomar el primer elemento
        if isinstance(creation_date, list):
            creation_date = creation_date[0]
        # Formatear la fecha como YYYY-MM-DD
        creation_date_str = creation_date.strftime('%Y-%m-%d')
        return creation_date_str
    except Exception as e:
        return "Error: No se pudo obtener la fecha de creación del TLD"



15., Muestre la fecha de creación para cada uno de los 13 dominios finales ¿Cuáles son los
dominios que podemos confirmar como sospechosos?

In [99]:
dga_expert_unique['creation_date'] = dga_expert_unique['domain_tld'].apply(get_creation_date)


suspect_domains = dga_expert_unique['domain_tld'][dga_expert_unique['creation_date'] == "Error: No se pudo obtener la fecha de creación del TLD"]
print("\nDominios sospechosos sin fecha de creación disponible:")
print(suspect_domains)


Dominios sospechosos sin fecha de creación disponible:
2                      aoltw.net
3                           home
8                 stayonline.net
9               22.110phpmyadmin
11               com.localdomain
13     110phpmyadmin.localdomain
15                 ntkrnlpa.info
23                           cn.
35                          wpad
36                       com.lan
37                       206.56"
45                       26-27.0
80                            FL
82                       saruman
83                       21.1201
91         whitecell.localdomain
97          vtlfccmfxlkgifuf.com
99                        22.254
100                      21-28.0
101         ejfodfmfxlkgifuf.xyz
102                      22.201:
Name: domain_tld, dtype: object


Recuerde que los dominios DGA son conocidos por formarse de forma aleatoria: secuencias
aleatorias de caracteres, no palabras. Indique que dominios sospechosos tienen este patrón y
que pueden confirmarse como dominios DGA.



Los siguientes dominios fueron identificados como **sospechosos** por cumplir una o más de las siguientes condiciones:

- No tienen fecha de creación disponible.
- Contienen cadenas aleatorias o estructuras no comunes en dominios legítimos.
- resentan combinaciones extrañas de números, subdominios o sufijos.

```text
- aoltw.net                  → no reconocido, no tiene información WHOIS.
- home                      → TLD atípico, muy genérico.
- stayonline.net            → nombre ambiguo, sin WHOIS válido.
- 22.110phpmyadmin          → formato numérico + referencia a herramienta técnica.
- com.localdomain           → no es un dominio válido a nivel público.
- 110phpmyadmin.localdomain → mezcla técnica sospechosa y TLD inválido.
- ntkrnlpa.info             → nombre muy técnico, relacionado con sistemas operativos (NT Kernel).
- cn.                       → dominio incompleto.
- wpad                      → potencial abuso del protocolo de autoconfiguración.
- com.lan                   → TLD inválido para internet pública.
- 206.56                   → estructura similar a una IP truncada.
- 26-27.0                   → nombre tipo IP, estructura inválida.
- FL                        → excesivamente corto, ambiguo, sin TLD.
- saruman                   → sin contexto, sin TLD, posible nombre falso.
- 21.1201                   → formato de número flotante, no tiene estructura de dominio.
- whitecell.localdomain     → dominio local ficticio.
- vtlfccmfxlkgifuf.com      → cadena aleatoria, típica de un DGA.
- 22.254                   → estructura incompleta tipo IP.
- 21-28.0                   → combinación numérica, no válida como dominio.
- ejfodfmfxlkgifuf.xyz      → cadena aleatoria + TLD legítimo, altamente sospechoso.
- 22.201:                  → notación mal formada (dos puntos al final), no es dominio válido.
